multicast/discovery_test.go

```diff
@@ -48,15 +93,31 @@ func (m *multicastSuite) Test_Lookup() {
 		// Use the loopback interface as it should always be there on any test system.
 		discovery := NewDiscovery("lo", 9444)
 
-		ctx, cancel := context.WithCancel(context.Background())
-		err := discovery.Respond(ctx, c.responseInfo)
+		err := discovery.Respond(context.Background(), c.responseInfo)
 		m.Require().NoError(err)
 
-		receivedInfo, err := discovery.Lookup(ctx, c.lookupVersion)
-		m.Require().NoError(err)
-		m.Require().Equal(&c.responseInfo, receivedInfo)
+		if c.modifier != nil {
+			c.modifier(discovery)
+		}
+
+		testDiscovery := NewDiscovery(c.lookupIface, c.lookupPort)
+
+		ctx := context.Background()
+		if c.lookupTimeout > 0 {
+			ctx, _ = context.WithTimeoutCause(ctx, c.lookupTimeout, fmt.Errorf("Timeout exceeded"))
```

Comment:
I don't know this new `context.WithTimeoutCause` function :) What is it exactly used for here? Why not using `context.WithTimeout` ?

Reply:
It helps to to reason about why a context got cancelled and allows you to transport the reason for cancellation to any recipient who is watching the context.

Take this as an example:

```go
ctx := context.WithTimeoutCause(ctx, 10*time.Second, fmt.Errorf("A more descriptive timeout error than the standard one")

// do something

go func() {
     <-ctx.Done()
     // Now the context is done but what has happened?
     // Was the parent context cancelled or our child one?
     err := context.Cause(ctx)
     fmt.Println(err.Error()) // Returns "A more descriptive timeout error than ..."
}()

```

So you can both override the default "context exceeded" error (e.g. there is also `WithCancelCause` that returns a canceller that allows you to put in an arbitrary `error` when cancelling) but you can also use it to react on certain errors with `errors.Is()` to see if your child context got cancelled with a specific error or if the parent context got cancelled with another error.

Reply:
In this PR its used to overwrite the default error.

Reply:
Got it :) Thanks!